doc/man/man4/momod.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/momod.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH MOMOD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B momod
\- Minimal OSI to ASCE/Presentation Module.
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.br
.B #include <ioctl.h>
.HP
.BI "int open(" xti_device ", " o_flags );
.HP
.BI "int ioctl(" xti_stream ", " I_PUSH ", " \*(lqmomod\*(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B momod
is a service interface conversion module that converts betweent the Minimal OSI XTI interface
.[
xns98
.]
described in
.BR xti_mosi (3),
and the ACSE/Presentation Library Interface service primitive interface described in
.BR apli (3).
The purpose of this module is to autopush the module over drivers implementing the
.BR xti_mosi (3)
interface for direct use by the
.BR ap_open (3)
function of the
.BR libapli (3)
or
.BR libxap (3)
libraries.
.PP
In particular, the
.BR lpp (4)
driver implements the Lightweight Presentation Protocol
.\"
.\"
.SH USAGE
.PP
The
.B momod
STREAMS module is not normally pushed directly by an XTI-mOSI
client programs, but is either pushed or autopushed as
necessary, either by the OSI network configuration daemon,
.BR osinetd (8),
or the autopush facility,
.BR autopush (8).
Nevertheless, the module may be pushed by clients of sufficient
privilege to acheive specific effects, such as for testing.
.\"
.\"
.SH INTERFACE
.PP
The upper service interface for the module is
the Transport Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
as interpreted for XTI-mOSI
.[
xns98
.]
operation,
.BR xti_mosi (3).
The lower service interface for the module is
the Transport Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
as interpreted for OSI
.[
xns98
.]
operation,
.BR xti_osi (3).
.\"
.\"
.SH ADDRESSES
.PP
Addresses consist of an AP-Title,
AE-Qualifier, and Presentation address.  The
Presentation Address (PSAP) consists of a
P-Selector and a TSAP.  The TSAP, in turn,
consists of a T-Selector and an NSAP.  The
NSAP is an OSI network address.
.PP
Addresses are formated as described in
.BR xti_mosi (3).
.\"
.\"
.SH OPTIONS
.PP
Options and option lists are internally
formated as described for XTI in
.BR t_optmgmt (3).
Option name and values are as described in
.BR xti_mosi (3).
.\"
.\"
.\".SH PRIMITIVES
.\".PP
.\"
.\"
.SH IOCTLS
.PP
.B momod
does not implement any input-output controls
of its own an passes all received
input-output controls downstream.
.B momod
is intended on being used with the
.BR timod (4)
module when used with the XTI Library,
.[
xti revision 1.0
.]
.BR xti (3).
.\"
.\"
.\".SH RETURN
.\".PP
.\"
.\"
.\".SH ERRORS
.\".PP
.\"
.\"
.\".SH NOTICES
.\".PP
.\"
.\"
.\".SH IMPLEMENTATION
.\".PP
.\"
.\"
.SH FILES
.PP
.TP
.B /lib/modules/striso/streams-momod.ko
provides the
.B momod
STREAMS pushable module named
.RI \(lq momod \(rq.
.\"
.\"
.SH MODULES
.PP
.RB \(lq momod \(rq
.I STREAMS
module.
.\"
.\"
.\".SH DEVICES
.\".PP
.\"
.\"
.SH SEE ALSO
.PP
.BR apli (3),
.BR ap_open (3),
.BR autopush (8).
.BR libapli (3),
.BR libxap (3),
.BR lpp (4),
.BR osinetd (8),
.BR test-mosi (8),
.BR timod (4),
.BR tpi (7),
.BR xti (3),
.BR xti_mosi (3).
.\"
.\"
.SH BUGS
.PP
.B xapmod
will be tested with the
.BR test-mosi (8)
test case executable and autotest conformance test suite.
.B momod
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
This
.B @PACKAGE_TITLE@
implementation of
.B momod
is compatible with
XTI-mOSI,
.[
xns98
.]
APLI,
.[
apli unix international
.]
XAP
.[
xap c303
.]
and descriptions for
.IR HP-UX \(rg,
.[
hpuxapli
.]
.IR Solaris \(rg,
.[
sunlink-apli
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
Other implementations do not document any cooperating module nor
input-output controls.  Portable applications should never use
input-output controls directly.
.IP \(em
Other implementations do not document primitive format between
the library and any cooperating module.
.IP \(em
Binary compatibility is not attempted.  Binary compatibility is
not possible due to the lack of standardized symbolic
constants.  Any binary compatibility experienced is a bug and
is scheduled to be removed on the next release.
.IP \(em
.IP \(em
.PP
See
.BR xti_mosi "(3),
.BR apli "(3),
.BR xap "(3),
.BR oopi "(4),
.BR apmod "(4),
.BR xapmod "(4) and
.BR STREAMS (4)
for additional compatibility information.
.PP
Compatibility will be tested using the
.BR test-mosi (8)
test case executable and the
.I @PACKAGE_TITLE@
.B autotest
test suite.
.\"
.\"
.SH CONFORMANCE
.PP
None.
.PP
Conformance will be tested using the
.BR test-mosi (8)
test case executable and the
.I @PACKAGE_TITLE@
.B autotest
test suite.
.\"
.\"
.SH HISTORY
.PP
The
.B momod
module is new for OpenSS7
.BR "striso version 0.9.2.4" .
.\"
.\"
.SH VERSIONS
.PP
This manual page was written for the OpenSS7 \fBmomod\fP(4) module.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/momod.c
/*****************************************************************************

 @(#) File: src/modules/momod.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/momod.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 * Minimal OSI (mOSI) Module
 *
 * The mosi(4) module provides a cooperating module for the XTI-mOSI interface
 * for the XTI Library (XTI).  The lower boundary service interface is the
 * Transport Provider Interface (TPI) and the upper boundary service interface
 * is the XTI-mOSI interface.  This module contains Session, Presentation and
 * ACSE protocol layers.
 */

#include <sys/os7/compat.h>

/*
 * These are for TPI definitions.
 */
#if defined HAVE_TIHDR_H
#   include <tihdr.h>
#else
#   include <sys/tihdr.h>
#endif

#define MOSI_DESCRIP	"SVR 4.2 MP RFC 1698 Minimal OSI (mOSI) STREAMS Module"
#define MOSI_EXTRA	"Part of the OpenSS7 OSI Stack for Linux Fast-STREAMS"
#define MOSI_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define MOSI_REVISION	"OpenSS7 src/modules/momod.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define MOSI_DEVICE	"SVR 4.2 MP STREAMS MOSI Module for RFC 1698 (mOSI)"
#define MOSI_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define MOSI_LICENSE	"GPL"
#define MOSI_BANNER	MOSI_DESCRIP	"\n" \
			MOSI_EXTRA	"\n" \
			MOSI_COPYRIGHT	"\n" \
			MOSI_REVISION	"\n" \
			MOSI_DEVICE	"\n" \
			MOSI_CONTACT
#define MOSI_SPLASH	MOSI_DEVICE	" - " \
			MOSI_REVISION

#ifdef LINUX
MODULE_AUTHOR(MOSI_CONTACT);
MODULE_DESCRIPTION(MOSI_DESCRIP);
MODULE_SUPPORTED_DEVICE(MOSI_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(MOSI_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-mosi");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#ifndef MOSI_MOD_NAME
#define MOSI_MOD_NAME		"mosi"
#endif				/* MOSI_MOD_NAME */
#ifndef MOSI_MOD_ID
#define MOSI_MOD_ID		0
#endif				/* MOSI_MOD_ID */

/*
 * STREAMS Definitions
 */

#define MOD_ID		MOSI_MOD_ID
#define MOD_NAME	MOSI_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	MOSI_BANNER
#else				/* MODULE */
#define MOD_BANNER	MOSI_SPLASH
#endif				/* MODULE */

static struct module_info mosi_minfo = {
	.mi_idnum = MOD_ID,		/* Module ID number */
	.mi_idname = MOD_NAME,		/* Module name */
	.mi_minpsz = 0,			/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,		/* Max packet size accepted */
	.mi_hiwat = STRHIGH,		/* Hi water mark */
	.mi_lowat = STRLOW,		/* Lo water mark */
};

static struct module_stat mosi_rstat __attribute__ ((aligned(SMP_CACHE_BYTES)));
static struct module_stat mosi_wstat __attribute__ ((aligned(SMP_CACHE_BYTES)));

static streamscall int mosi_open(queue_t *, dev_t *, int, int, cred_t *);
static streamscall int mosi_close(queue_t *, int, cred_t *);

static streamscall int mosi_rput(queue_t *, mblk_t *);
static streamscall int mosi_wput(queue_t *, mblk_t *);

static struct qinit mosi_rinit = {
	.qi_putp = mosi_rput,		/* Read put (message from below) */
	.qi_qopen = &mosi_open,		/* Each open */
	.qi_qclose = &mosi_close,	/* Last close */
	.qi_minfo = &mosi_minfo,	/* Module information */
	.qi_mstat = &mosi_rstat,	/* Module statistics */
};

static struct qinit mosi_winit = {
	.qi_putp = mosi_wput,		/* Write put (message form above) */
	.qi_minfo = &mosi_minfo,	/* Module information */
	.qi_mstat = &mosi_wstat,	/* Module statistics */
};

struct streamtab mosi_info = {
	.st_rdinit = &mosi_rinit,
	.st_wrinit = &mosi_winit,
};

/*
 * PUTP and SRVP, put and service procedures
 */

/*
 * Read-Side Processing
 */

/**
 * mosi_r_proto - process M_(PC)PROTO message block on read side
 * @q: read queue
 * @mp: the message
 *
 * Note that much of this processing is stateless and pretty automatic.  When
 * a request is generated on the write side, we transform to a transport TPI
 * message, pass it down and wait for the response.  Here the response is most
 * often simply translated into the corresponing mOSI response.
 */
static streams_fastcall streams_inline __hot_get int
mosi_r_proto(queue_t *q, mblk_t *mp)
{
	switch (*(t_scalar_t) mp->b_rptr) {
	case T_CONN_IND:
	case T_CONN_CON:
	case T_DISCON_IND:
	case T_DATA_IND:
	case T_EXDATA_IND:
	case T_INFO_ACK:
	case T_BIND_ACK:
	case T_ERROR_ACK:
	case T_OK_ACK:
	case T_UNITDATA_IND:
	case T_UDERROR_IND:
	case T_OPTMGMT_ACK:
	case T_ORDREL_IND:
	case T_OPTDATA_IND:
	case T_ADDR_ACK:
	case T_CAPABILITY_ACK:
/*
 * None of the following messages should appear at the read-side put
 * procedure.  Only upward defined TPI messages should appear at the read-side
 * put procedure.  Discard these.
 */
	case T_CONN_REQ:
	case T_CONN_RES:
	case T_DISCON_REQ:
	case T_DATA_REQ:
	case T_EXDATA_REQ:
	case T_INFO_REQ:
	case T_BIND_REQ:
	case T_UNBIND_REQ:
	case T_UNITDATA_REQ:
	case T_OPTMGMT_REQ:
	case T_ORDREL_REQ:
	case T_OPTDATA_REQ:
	case T_ADDR_REQ:
	case T_CAPABILITY_REQ:
		__swerr();
	default:
		freemsg(mp);
		return (0);
	}
}

static streamscall int
mosi_r_flush(queue_t *q, mblk_t *mp)
{
	if (mp->b_rptr[0] & FLUSHR) {
		if (mp->b_rptr[0] & FLUSHBAND)
			flushband(q, mp->b_rptr[1], FLUSHDATA);
		else
			flushq(q, FLUSHDATA);
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
mosi_r_data(queue_t *q, mblk_t *mp)
{
	__swerr();
	return (EFAULT);
}

static streams_fastcall streams_noinline int
mosi_r_msg(queue_t *q, mblk_t *mp)
{
	switch (mp->b_datap->db_type) {
	case M_PROTO:
	case M_PCPROTO:
		return mosi_r_proto(q, mp);
	case M_DATA:
		return mosi_r_data(q, mp);
	case M_FLUSH:
		return mosi_r_flush(q, mp);
	}
	putnext(q, mp);
	return (0);
}

static streamscall __hot_get int
mosi_rput(queue_t *q, mblk_t *mp)
{
	if (mp->b_datap->db_type == M_PROTO)
		return mosi_r_proto(q, mp);
	return mosi_r_msg(q, mp);
}

/*
 * Write-Side Processing
 */

static streams_fastcall streams_inline __hot_put int
mosi_w_proto(queue_t *q, mblk_t *mp)
{
	switch (*(t_scalar_t *) mp->b_rptr) {
	case T_CONN_REQ:
	case T_CONN_RES:
	case T_DISCON_REQ:
	case T_DATA_REQ:
	case T_EXDATA_REQ:
	case T_INFO_REQ:
	case T_BIND_REQ:
	case T_UNBIND_REQ:
	case T_UNITDATA_REQ:
	case T_OPTMGMT_REQ:
	case T_ORDREL_REQ:
	case T_OPTDATA_REQ:
	case T_ADDR_REQ:
	case T_CAPABILITY_REQ:
/*
 * None of the following messages should appear at the write-side put
 * procedure.  Only downward deinfed TPI messages should appear at the
 * write-side put procedure.  Discard these.
 */
	case T_CONN_IND:
	case T_CONN_CON:
	case T_DISCON_IND:
	case T_DATA_IND:
	case T_EXDATA_IND:
	case T_INFO_ACK:
	case T_BIND_ACK:
	case T_ERROR_ACK:
	case T_OK_ACK:
	case T_UNITDATA_IND:
	case T_UDERROR_IND:
	case T_OPTMGMT_ACK:
	case T_ORDREL_IND:
	case T_OPTDATA_IND:
	case T_ADDR_ACK:
	case T_CAPABILITY_ACK:
		__swerr();
	default:
		freemsg(mp);
		return (0);
	}
}

static streamscall int
mosi_w_flush(queue_t *q, mblk_t *mp)
{
	if (mp->b_rptr[0] & FLUSHW) {
		if (mp->b_rptr[0] & FLUSHBAND)
			flushband(q, mp->b_rptr[1], FLUSHDATA);
		else
			flushq(q, FLUSHDATA);
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
mosi_w_data(queue_t *q, mblk_t *mp)
{
	/* FIXME: this should be ok. */
	__swerr();
	return (EFAULT);
}

static streams_fastcall streams_noinline int
mosi_w_msg(queue_t *q, mblk_t *mp)
{
	switch (mp->b_datap->db_type) {
	case M_PROTO:
	case M_PCPROTO:
		return mosi_w_proto(q, mp);
	case M_DATA:
		return mosi_w_data(q, mp);
	case M_FLUSH:
		return mosi_w_flush(q, mp);
	}
	putnext(q, mp);
	return (0);
}

/**
 * mosi_wput - write-side put procedure
 * @q: write queue
 * @mp: message to put
 *
 * Messages arriving at the write-side put procedure are messages from above
 * that are TPI formatted messages.  For all intents and purposes, the module
 * above us should believe that we are any mOSI based XTI provider.
 */
statis streamscall __hot_put int
mosi_wput(queue_t *q, mblk_t *mp)
{
	int type = mp->b_datap->db_type;

	/* Fast path. */
	if (type == M_PROTO)
		return mosi_w_proto(q, mp);
	return mosi_w_msg(q, mp);
}

static streamscall int
mosi_open(queue_t *q, dev_t *devp, int oflags, int sflag, cred_t *crp)
{
	return (ENXIO);
}

static streamscall int
mosi_close(queue_t *q, int oflags, cred_t *crp)
{
	return (ENXIO);
}

/*
 * --------------------------------------------------------------------------
 *
 *  QOPEN and QCLOSE routines
 *
 * --------------------------------------------------------------------------
 */

static caddr_t mosi_opens = NULL;

static streamscall int
mosi_qopen(queue_t *q, dev_t *devp, int oflags, int sflag, cred_t *crp)
{
	struct priv *p;
	mblk_t *mp;
	int err;

	if (q->q_ptr)
		return (0);	/* already open */
	if (!(mp = allocb(sizeof(struct T_info_req), BPRI_WAITOK)))
		return (ENOBUFS);
	if ((err = mi_open_comm(&mosi_opens, sizeof(*p), q, devp, oflags, sflag, crp))) {
		freeb(mp);
		return (err);
	}
	p = PRIV(q);

	/* intiialize private structure */
	p->r_priv.priv = p;
	p->r_priv.mo = &p->w_priv;
	p->r_priv.oq = WR(q);
	p->r_priv.state = TS_UNBND;
	p->r_priv.oldstate = TS_UNBND;

	p->w_priv.priv = p;
	p->w_priv.tp = &p->r_priv;
	p->w_priv.oq = q;
	p->w_priv.state = TS_UNBND;
	p->w_priv.oldstate = TS_UNBND;

	/* issue an immediate information request */
	DB_TYPE(mp) = M_PCPROTO;
	((struct T_info_req *) mp->b_wptr)->PRIM_type = T_INFO_REQ;
	mp->b_wptr += sizeof(struct T_info_req);

	qprocson(q);
	putnext(q, mp);
	return (0);
}

static streamscall int
mosi_qclose(queue_t *q, int oflags, cred_t *crp)
{
	struct priv *p = PRIV(q);
	(void)p;
	qprocsoff(q);
	/* FIXME: free timers and other things */
	mi_close_comm(&mosi_opens, q);
	return (0);
}

/*
 * --------------------------------------------------------------------------
 *
 *  STREAMS INITIALIZATION
 *
 * --------------------------------------------------------------------------
 */

static struct qinit mosi_rinit = {
	.qi_putp = tp_rput,		/* Read put (message from below) */
	.qi_srvp = tp_rsrv,		/* Read service */
	.qi_qopen = &mosi_open,		/* Each open */
	.qi_qclose = &mosi_close,	/* Last close */
	.qi_minfo = &mosi_minfo,	/* Module information */
	.qi_mstat = &mosi_rstat,	/* Module statistics */
};

static struct qinit mosi_winit = {
	.qi_putp = mo_wput,		/* Write put (message form above) */
	.qi_srvp = mo_wsrv,		/* Write service */
	.qi_minfo = &mosi_minfo,	/* Module information */
	.qi_mstat = &mosi_wstat,	/* Module statistics */
};

struct streamtab mosi_info = {
	.st_rdinit = &mosi_rinit,   /* Upper RD queue */
	.st_wrinit = &mosi_winit,   /* Upper WR queue */
};

static modID_t modid = MOD_ID;

/*
 * --------------------------------------------------------------------------
 *
 *  STREAMS REGISTRATION
 *
 * --------------------------------------------------------------------------
 */

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for mOSI. (0 for allocation.)");

static struct fmodsw mosi_fmod = {
	.f_name = MOD_NAME,
	.f_str = &mosi_modinfo,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

/*
 * --------------------------------------------------------------------------
 *
 *  LINUX KERNEL MODULE INITIALIZATION
 *
 * --------------------------------------------------------------------------
 */

static __init int
mosi_modinit(void)
{
	int err;

	cmn_err(CE_NOTE, MOD_BANNER);
	if ((err = register_strmod(&mosi_fmod)) < 0) {
		cmn_err(CE_WARN, "%s: could not register module %d", MOD_NAME, (int) modid);
		return (err);
	}
	if (modid == 0)
		modid = err;
	return (0);
}

static __exit void
mosi_modexit(void)
{
	int err;

	if ((err = unregister_strmod(&mosi_fmod)) < 0) {
		cmd_err(CE_WARN, "%s: could not unregister module, err = %d", MOD_NAME, err);
		return;
	}
	return;
}

module_init(mosi_modinit);
module_exit(mosi_modexit);
